src/value_generators/include/value_generators/DocumentGenerator.hpp

```diff
@@ -176,15 +178,31 @@ struct IntegerValueType {
     constexpr static ValueType valueType() {
         return ValueType::Integer;
     }
-    constexpr static OutputType convert(const Value& value) {
+    static OutputType convert(const Value& value) {
         auto out = value.tryAsInt64();
         if (!out) {
-            throw InvalidValueGeneratorSyntax("Unknown");
+            std::stringstream error;
+            error << "Expected integer, but got " << value;
+            throw InvalidValueGeneratorSyntax(error.str());
         }
         return *out;
     }
 };
 
+struct DoubleValueType {
+    using OutputType = double;
+    constexpr static ValueType valueType() {
+        return ValueType::Double;
+    }
+    static OutputType convert(const Value& value) {
+        auto maybeInt = value.tryAsInt64();
```

Comment:
Ya this is somewhat temporary. I'm going to add a `DoubleGenerator` type and the `ConstantDoubleGenerator` impl will do `Node::as<double>()`.

> not all int64_ts can be exactly represented...

IMHO I'd leave this up to the built-in behavior of `Node::as<double>()`. If it encounters something that can't be converted (e.g. an unreasonably large number of digits) it throws a `bad conversion` exception.

And fwiw on my machine `std::numeric_limits<double>::max()` reports as `1.79769e+308`.

Reply:
> IMHO I'd leave this up to the built-in behavior of `Node::as<double>()`. If it encounters something that can't be converted (e.g. an unreasonably large number of digits) it throws a `bad conversion` exception.

But you aren't doing that. You're having yaml-cpp convert the value to an `int64_t` and then asking C++ to convert that value to a `double`. C++ isn't going to do you any favors and tell you whether that conversion incurs a loss of precision.

> And fwiw on my machine `std::numeric_limits<double>::max()` reports as `1.79769e+308`.

I'm not sure how that's relevant. [`Number.MAX_SAFE_INTEGER`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) is the relevant value and isn't going to different between machines.

Reply:
Is there a reason why reverting to say `const double _p` in `BinomialIntExpression`, `NegativeBinomialIntExpression`, and `GeometricIntExpression`, as well as `const double _mean` in `PoissonIntExpression` is undesirable?

Reply:
`Number.MAX_SAFE_INTEGER` is for javascript, ya? Fwiw I was able to get the cpp yaml parser to spit out a `long long` that was over 100 digits long in base10.

But I think this whole thing is all for naught: I can't think of a reason why we'd want to read yaml-borne values as int32 and not just always use int64. Users don't explicitly specify a type so we just implicitly try it as a 32 or a 64 if necessary. It seems less surprising to always use int64, and it makes downstream concerns a lot easier (especially in testing) to be able to assume that all numbers are either `double`s or `int64`s. Please push back if you disagree - but I may punt supporting `int32`s to a separate ticket where we can figure out a more explicit syntax (maybe similar to how the shell and friends do `{$numberLong:}` etc?)

Reply:
> Is there a reason why reverting to say const double _p in BinomialIntExpression, NegativeBinomialIntExpression, and GeometricIntExpression, as well as const double _mean in PoissonIntExpression is undesirable?

As opposed to the change to use a generator for the values? It feels "right" to have everything be applied as a template similar to how `RandomString.length` is templated etc. Plus I think it'll make later changes clearer to have everything be consistently a typed-generator.

Reply:
> `Number.MAX_SAFE_INTEGER` is for javascript, ya?

It's all the same IEEE 754 floating-point value representation.

> Fwiw I was able to get the cpp yaml parser to spit out a long long that was over 100 digits long in base10.

That doesn't make any sense to me because `log(1e100) / log(2) ~= 332`. There aren't enough bits in a 64-bit integral value.

> I can't think of a reason why we'd want to read yaml-borne values as int32 and not just always use int64.

BSON has both `int32_t` and `int64_t` types, which may have different performance characteristics given their different widths. I agree about eventually needing to offer syntax to explicitly specify `int64_t` (e.g. an 'L' suffix) but I don't see why we're throwing away support for `int32_t`.

> As opposed to the change to use a generator for the values? It feels "right" to have everything be applied as a template similar to how `RandomString.length` is templated etc. Plus I think it'll make later changes clearer to have everything be consistently a typed-generator.

Until there is a non-`ConstantExpression` way to generate `double` values, I don't think it is worth it adding the extra layer. We can defer it until we have such an expression and are forced to have a better story around conversions between numeric types.

Reply:
> I don't see why we're throwing away support for int32_t.

We narrow values into int32 if they fit otherwise we leave them as int32s. This is probably not what users expect - I would expect the output type of the DocumentGenerator to always be the same not "the smallest necessary type that could fit the randomly-generated value". E.g. it is there's no way to explicitly generate an int64 with value 0.

I've reverted the change to kill int32 but I think it may come back to haunt us later. If we want value-generators to be strictly typesafe we'd need to introduce int32 and int64 variants of the `RandomIntExpression` classes.

I also got rid of trying to treat Doubles as int64s. I'll add back the `const double`s.

Reply:
> We narrow values into int32 if they fit otherwise we leave them as int32s. This is probably not what users expect - I would expect the output type of the DocumentGenerator to always be the same not "the smallest necessary type that could fit the randomly-generated value".

There aren't any randomly generated `int32_t`s - there are only randomly generated `int64_t`s and `std::string`s. `int32_t`s and `double`s can only ever appear as `ConstantExpression`s. I couldn't really speak to what users would expect in this context because there's no explicit syntax (everything would have been `double`s if this were JavaScript). It may not even matter in practice because the server is going widen the type from an `int32_t` to an `int64_t` automatically.

```
> db.mycoll.insert({_id: 0, a: NumberInt(Math.pow(2, 20))})
> db.mycoll.update({_id: 0}, {$mul: {a: NumberInt(Math.pow(2, 20))}})
> db.mycoll.findOne({_id: 0})
{ "_id" : 0, "a" : NumberLong("1099511627776") }
```

What I meant when I said "I don't see why we're throwing away support for `int32_t`" is "I don't see why this pull request would be throwing away support for `int32_t` when its purpose is tackling the syntax for generating values."

> E.g. it is there's no way to explicitly generate an int64 with value 0.

The syntax is laughable awful but you can do

```
{^RandomInt: {min: 0, max: 0}}
```

> I also got rid of trying to treat Doubles as int64s. I'll add back the `const double`s.

Thanks, just ping me again to review when you've made those changes.

Reply:
And we're back to `const double`s. Ready for another look whenever you can.